TheJumpCloud/support
# SA-2541: Get-JCResults Parallel Implementation

## Issues
* [SA-2541](https://jumpcloud.atlassian.net/browse/SA-2541) - Get-JCResults Parallel Implementation

## What does this solve?
This is laying the groundwork for allowing customers to utilize PowerShell 7 parallel processing when using specific Get functions. The Get-JCResults private function is intended to replace existing while loops to make API calls and return the results. A new boolean parameter, `-parallel $true/$false`, was introduced to determine if the API calls will be made in parallel or sequential. 

The next step is going to be removing the while loops in favor of this new function and introduce a `-parallel $true/$false` parameter in applicable functions

## Is there anything particularly tricky?
Parallel processing can only be utilized on PowerShell versions 7+. There are checks in place to determine if the current session has a compatible version as well as the `-parallel` parameter set to `true`.

if the version is not supported and `-parallel $true`, a warning will be displayed informing the user of the compatibility issue and the calls will be made sequentially

## How should this be tested?
Test to verify that both $rawResults contain the same information
```powershell
Import-Module "/Users/USERNAME/Documents/GitHub/support/PowerShell/JumpCloud Module/JumpCloud.psd1" -force
Import-Module "/Users/USERNAME/Documents/GitHub/support/PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCUserAgent.ps1" -force
Import-Module "/Users/USERNAME/Documents/GitHub/support/PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCResults.ps1" -force
Connect-JCOnline APIKEY
$JCUrlBasePath = 'https://console.jumpcloud.com'
$Group_ID = 'GROUPID'
$limitURL = "{0}/api/v2/usergroups/{1}/members" -f $JCUrlBasePath, $Group_ID
$rawResults = Get-JCResults -Url $limitURL -method "GET" -limit 100 -parallel $true -debug
$rawResults = Get-JCResults -Url $limitURL -method "GET" -limit 100 -parallel $false -debug
```
Measure time to complete for both functions (parallel *should* be faster, but not guaranteed) 
```powershell
Measure-Command { Get-JCResults -Url $limitURL -method "GET" -limit 100 -parallel $false }
Measure-Command { Get-JCResults -Url $limitURL -method "GET" -limit 100 -parallel $true }
```
